pythia/utils/text_utils.py

```diff
@@ -271,21 +261,93 @@ def search(self, t, data, scores):
             captions = torch.FloatTensor(self.complete_seqs[i]).unsqueeze(0)
         return captions
+
+class NucleusSampling:
+    def __init__(self, vocab, threshold=0.8):
+        self.vocab = vocab
+        self.vocab_size = vocab.get_size()
+        # Threshold for sum of probability
+        self.threshold = threshold
+        # Lists to store completed sequence
+        self.complete_seq = []
+
+    def init_batch(self, sample_list):
+        setattr(
+            self,
+            "seq",
+            sample_list.answers.new_full(
+                (1, 1), self.vocab.SOS_INDEX, dtype=torch.long
+            ),
+        )
+        # Add a dim and duplicate the tensor beam_size times across that dim
+        sample_list.image_feature_0 = (
+            sample_list.image_feature_0.unsqueeze(1)
+            .expand(-1, 1, -1, -1)
+            .squeeze(0)
+        )
+        return sample_list
+
+    def sample(self, t, data, scores):
+        # Convert scores to probabilities
+        scores = torch.nn.functional.softmax(scores, dim=1)
+        # Sort scores in descending order and then select the top m elements having sum more than threshold.
+        # We get the top_m_scores and their indices top_m_words
+        if t == 0:
+            top_m_scores, top_m_words = scores[0].sort(
+                0, True
+            )
+        else:
+            top_m_scores, top_m_words = scores.view(-1).sort(
+                0, True
+            )
+
+        lastIndex = 0
+        scoreSum = 0
+        for score in top_m_scores:
+            lastIndex += 1
+            scoreSum += score
+            if scoreSum >= self.threshold:
+                break
+
+        top_m_scores = torch.div(top_m_scores[:lastIndex], scoreSum)
+        top_m_words = top_m_words[:lastIndex]
+        # Zero value inside prev_word_inds because we are predicting a single stream of output.
+        prev_word_ind = torch.tensor([0])
+        # Get next word based on probabilities of top m words.
+        next_word_ind = top_m_words[torch.multinomial(top_m_scores, 1)]
+        # Add next word to sequence
+        self.seq = torch.cat(
+            [self.seq[prev_word_ind], next_word_ind.unsqueeze(1)], dim=1
+        )
+        # Check if sequence is complete
+        complete_inds, incomplete_inds = find_complete_inds(self, next_word_ind)
+
+        # If sequence is complete then return
+        if len(complete_inds) > 0:
+            self.complete_seq.extend(self.seq[complete_inds].tolist())
+            return True, data, 0
+
+        self.seq = self.seq[incomplete_inds]
+
+        # TODO: Make the data update generic for any type of model
+        # This is specific to BUTD model only.
+        data = update_data(data, prev_word_ind, next_word_ind, incomplete_inds)
+
+        return False, data, 1
+
+    def get_caption(self):
+        captions = torch.FloatTensor(self.complete_seq[0]).unsqueeze(0)
```

Comment:
This needs to be on proper device instead of cpu until unless the last moment where it will be decoded.

Reply:
I'm not sure what exactly is required here. Isn't decoding already complete when line 352 is executed?

Reply:
@apsdehal should this be considered resolved?